File: lib/schemas.ts
import { z } from 'zod';

export const ContactSchema = z.object({
  name: z.string().default(''),
  email: z.string().default(''),
  phone: z.string().default(''),
  location: z.string().default(''),
  linkedin: z.string().optional(),
});

export const WorkExperienceSchema = z.object({
  company: z.string().default(''),
  title: z.string().default(''),
  startDate: z.string().default(''),
  endDate: z.string().default(''),
  bullets: z.array(z.string()).default([]),
});

export const EducationSchema = z.object({
  institution: z.string().default(''),
  degree: z.string().default(''),
  field: z.string().default(''),
  year: z.string().default(''),
  cgpa: z.string().optional(),
});

export const ParsedProjectSchema = z.object({
  name: z.string().default(''),
  description: z.string().default(''),
  url: z.string().optional(),
  startDate: z.string().optional(),
  endDate: z.string().optional(),
});

export const ParsedLanguageSchema = z.object({
  name: z.string().default(''),
  proficiency: z.string().optional(),
});

export const ParsedResumeSchema = z.object({
  contact: ContactSchema,
  summary: z.string().default(''),
  experience: z.array(WorkExperienceSchema).default([]),
  education: z.array(EducationSchema).default([]),
  skills: z.array(z.string()).default([]),
  certifications: z.array(z.string()).default([]),
  noticePeriod: z.string().optional(),
  ctc: z.string().optional(),
  hasMultiColumn: z.boolean().default(false),
  hasTables: z.boolean().default(false),
  hasImages: z.boolean().default(false),
  rawText: z.string().optional(),
  projects: z.array(ParsedProjectSchema).optional(),
  languages: z.array(ParsedLanguageSchema).optional(),
});

export const ParsedJDSchema = z.object({
  jobTitle: z.string().default(''),
  requiredSkills: z.array(z.string()).default([]),
  preferredSkills: z.array(z.string()).default([]),
  yearsRequired: z.number().default(0),
  tools: z.array(z.string()).default([]),
  industryKeywords: z.array(z.string()).default([]),
  seniorityLevel: z.string().default(''),
  domain: z.string().default(''),
});

export const ScoreBreakdownSchema = z.object({
  keyword: z.number().default(0),
  formatting: z.number().default(0),
  naukri: z.number().default(0),
  content: z.number().default(0),
});

export const ATSScoreSchema = z.object({
  total: z.number().default(0),
  breakdown: ScoreBreakdownSchema,
  missingKeywords: z.array(z.string()).default([]),
  matchedKeywords: z.array(z.string()).default([]),
  formattingIssues: z.array(z.string()).default([]),
  naukriIssues: z.array(z.string()).default([]),
  contentIssues: z.array(z.string()).default([]),
  topFixes: z.array(z.string()).default([]),
});

export const RewrittenResumeSchema = z.object({
  contact: ContactSchema,
  summary: z.string().default(''),
  experience: z.array(WorkExperienceSchema).default([]),
  education: z.array(EducationSchema).default([]),
  skills: z.array(z.string()).default([]),
  certifications: z.array(z.string()).default([]),
  naukriProfileText: z.string().default(''),
  noticePeriod: z.string().optional(),
  ctc: z.string().optional(),
});

export const RewriteResultSchema = z.object({
  original: ParsedResumeSchema,
  rewritten: RewrittenResumeSchema,
});

export const NaukriProfileSchema = z.object({
  headline: z.string().default(''),
  summary: z.string().default(''),
  keySkills: z.array(z.string()).default([]),
});

// --- New Resume Builder Types ---

export const ResumeContactSchema = z.object({
  firstName: z.string().default(''),
  lastName: z.string().default(''),
  jobTitle: z.string().default(''),
  email: z.string().default(''),
  phone: z.string().default(''),
  city: z.string().default(''),
  country: z.string().default(''),
  linkedin: z.string().default(''),
  github: z.string().default(''),
  website: z.string().default(''),
});

export const ResumeExperienceSchema = z.object({
  id: z.string().default(''),
  company: z.string().default(''),
  position: z.string().default(''),
  location: z.string().default(''),
  startDate: z.string().default(''),
  endDate: z.string().default(''),
  currentlyWorking: z.boolean().default(false),
  description: z.string().default(''),
});

export const ResumeEducationSchema = z.object({
  id: z.string().default(''),
  institution: z.string().default(''),
  degree: z.string().default(''),
  field: z.string().default(''),
  location: z.string().default(''),
  startDate: z.string().default(''),
  endDate: z.string().default(''),
  currentlyStudying: z.boolean().default(false),
  score: z.string().default(''),
});

export const ResumeSkillSchema = z.object({
  id: z.string().default(''),
  name: z.string().default(''),
  level: z.string().default('Intermediate'),
});

export const ResumeProjectSchema = z.object({
  id: z.string().default(''),
  name: z.string().default(''),
  description: z.string().default(''),
  url: z.string().default(''),
  startDate: z.string().default(''),
  endDate: z.string().default(''),
});

export const ResumeCertificationSchema = z.object({
  id: z.string().default(''),
  name: z.string().default(''),
  issuer: z.string().default(''),
  issueDate: z.string().default(''),
  expiryDate: z.string().default(''),
  doesNotExpire: z.boolean().default(true),
  credentialId: z.string().default(''),
  credentialUrl: z.string().default(''),
});

export const ResumeLanguageSchema = z.object({
  id: z.string().default(''),
  name: z.string().default(''),
  proficiency: z.string().default('Professional Working Proficiency'),
});

export const ResumeCustomItemSchema = z.object({
  id: z.string().default(''),
  name: z.string().default(''),
  description: z.string().default(''),
});

export const ResumeCustomSectionSchema = z.object({
  id: z.string().default(''),
  title: z.string().default(''),
  items: z.array(ResumeCustomItemSchema).default([]),
});

export const ResumePublicationSchema = z.object({
  id: z.string().default(''),
  title: z.string().default(''),
  publisher: z.string().default(''),
  date: z.string().default(''),
  coAuthors: z.string().default(''),
  url: z.string().default(''),
});

export const ResumeCourseSchema = z.object({
  id: z.string().default(''),
  name: z.string().default(''),
  platform: z.string().default(''),
  completionDate: z.string().default(''),
  certificateUrl: z.string().default(''),
});

export const ResumeAwardSchema = z.object({
  id: z.string().default(''),
  name: z.string().default(''),
  issuer: z.string().default(''),
  date: z.string().default(''),
  description: z.string().default(''),
});

export const ResumeVolunteerSchema = z.object({
  id: z.string().default(''),
  organization: z.string().default(''),
  role: z.string().default(''),
  location: z.string().default(''),
  startDate: z.string().default(''),
  endDate: z.string().default(''),
  currentlyVolunteering: z.boolean().default(false),
  description: z.string().default(''),
});

export const ResumeDataSchema = z.object({
  contact: ResumeContactSchema,
  summary: z.string().default(''),
  experience: z.array(ResumeExperienceSchema).default([]),
  education: z.array(ResumeEducationSchema).default([]),
  skills: z.array(ResumeSkillSchema).default([]),
  projects: z.array(ResumeProjectSchema).default([]),
  certifications: z.array(ResumeCertificationSchema).default([]),
  languages: z.array(ResumeLanguageSchema).default([]),
  publications: z.array(ResumePublicationSchema).default([]),
  courses: z.array(ResumeCourseSchema).default([]),
  awards: z.array(ResumeAwardSchema).default([]),
  volunteer: z.array(ResumeVolunteerSchema).default([]),
  customSections: z.array(ResumeCustomSectionSchema).default([]),
});

File: services/ats_scorer/schemas.py
"""
Pydantic schemas for the ATS scorer service.

Field names use camelCase throughout to exactly match the TypeScript interfaces
in lib/types.ts — this ensures the JSON payload needs zero transformation on
either the frontend or the Python side.
"""
from __future__ import annotations

from pydantic import BaseModel, Field


# ── Inbound resume payload (mirrors ResumeData in lib/types.ts) ───────────────

class ResumeContactPayload(BaseModel):
    firstName: str = ""
    lastName: str = ""
    jobTitle: str = ""
    email: str = ""
    phone: str = ""
    city: str = ""
    country: str = ""
    linkedin: str = ""
    github: str = ""
    website: str = ""


class ResumeExperiencePayload(BaseModel):
    id: str = ""
    company: str = ""
    position: str = ""
    location: str = ""
    startDate: str = ""
    endDate: str = ""
    currentlyWorking: bool = False
    description: str = ""  # newline-delimited bullet points


class ResumeEducationPayload(BaseModel):
    id: str = ""
    institution: str = ""
    degree: str = ""
    field: str = ""
    location: str = ""
    startDate: str = ""
    endDate: str = ""
    currentlyStudying: bool = False
    score: str = ""


class ResumeSkillPayload(BaseModel):
    id: str = ""
    name: str = ""
    level: str = ""


class ResumeProjectPayload(BaseModel):
    id: str = ""
    name: str = ""
    description: str = ""
    url: str = ""
    startDate: str = ""
    endDate: str = ""


class ResumeCertificationPayload(BaseModel):
    id: str = ""
    name: str = ""
    issuer: str = ""
    issueDate: str = ""
    expiryDate: str = ""
    doesNotExpire: bool = False
    credentialId: str = ""
    credentialUrl: str = ""


class ResumeDataPayload(BaseModel):
    contact: ResumeContactPayload = Field(default_factory=ResumeContactPayload)
    summary: str = ""
    experience: list[ResumeExperiencePayload] = []
    education: list[ResumeEducationPayload] = []
    skills: list[ResumeSkillPayload] = []
    projects: list[ResumeProjectPayload] = []
    certifications: list[ResumeCertificationPayload] = []


# ── Request ───────────────────────────────────────────────────────────────────

class ATSScoreRequest(BaseModel):
    resume: ResumeDataPayload = Field(default_factory=ResumeDataPayload)
    jdText: str = ""  # camelCase matches frontend useUIStore field name


# ── Response (must match ATSScore in lib/types.ts exactly) ───────────────────

class ScoreBreakdown(BaseModel):
    """Mirrors TypeScript ScoreBreakdown interface."""
    keyword: int
    formatting: int
    naukri: int
    content: int


class ATSScoreResponse(BaseModel):
    """
    Mirrors TypeScript ATSScore interface exactly.
    Extra fields (semanticSimilarity, extractedSkills) are safe addenda —
    TypeScript will ignore unknown JSON keys when cast with 'as ATSScore'.
    """
    total: int
    breakdown: ScoreBreakdown
    missingKeywords: list[str]
    matchedKeywords: list[str]
    formattingIssues: list[str]
    naukriIssues: list[str]
    contentIssues: list[str]
    topFixes: list[str]
    # Extended fields — useful for the frontend keyword scanner
    semanticSimilarity: float = 0.0   # raw cosine similarity 0.0–1.0
    extractedSkills: list[str] = []   # skills extracted from the JD by NLP
